File: Codes/Prims.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <climits>

using namespace std;

// Edge structure for graph representation
struct Edge {
    int src, dest, weight;
};

// Comparator for priority queue to store edges in Prim's algorithm
struct Compare {
    bool operator()(const Edge& a, const Edge& b) {
        return a.weight > b.weight;  // Min-heap based on edge weight
    }
};

// Function to compute MST using Prim's algorithm
vector<Edge> primMST(vector<vector<pair<int, int>>>& graph, int n) {
    // Min-heap to select minimum weight edge at each step
    priority_queue<Edge, vector<Edge>, Compare> pq;
    vector<bool> inMST(n, false);  // Track vertices included in MST
    vector<Edge> mst;  // Store the resulting MST

    // Start with vertex 0
    inMST[0] = true;
    for (const auto& neighbor : graph[0]) {
        pq.push({0, neighbor.first, neighbor.second});
    }

    while (!pq.empty() && mst.size() < n - 1) {
        Edge minEdge = pq.top();
        pq.pop();

        // If destination vertex is already in MST, skip this edge
        if (inMST[minEdge.dest]) continue;

        // Add edge to MST
        mst.push_back(minEdge);
        inMST[minEdge.dest] = true;

        // Add all edges from the newly added vertex to the priority queue
        for (const auto& neighbor : graph[minEdge.dest]) {
            if (!inMST[neighbor.first]) {
                pq.push({minEdge.dest, neighbor.first, neighbor.second});
            }
        }
    }

    return mst;
}

// Function to add redundancy edges to meet reliability constraints
vector<Edge> addRedundancy(vector<Edge>& edges, vector<Edge>& mst, int n, int k) {
    vector<Edge> redundantEdges;
    vector<vector<int>> mstGraph(n, vector<int>(n, INT_MAX));
    
    for (const auto& edge : mst) {
        mstGraph[edge.src][edge.dest] = edge.weight;
        mstGraph[edge.dest][edge.src] = edge.weight;
    }
    
    for (const auto& edge : edges) {
        if (mstGraph[edge.src][edge.dest] == INT_MAX) {
            redundantEdges.push_back(edge);
        }
    }
    
    // To keep it simple, just pick the first k redundant edges
    if (redundantEdges.size() > k) {
        redundantEdges.resize(k);
    }
    
    return redundantEdges;
}

// Function to validate edge weights within a specified range
void validateEdgeWeights(const vector<Edge>& edges, int minWeight, int maxWeight) {
    for (const auto& edge : edges) {
        if (edge.weight < minWeight || edge.weight > maxWeight) {
            throw invalid_argument("Edge weight out of range: " + to_string(edge.weight));
        }
    }
}

// Function to calculate the total weight of the MST
int calculateMSTWeight(const vector<Edge>& mst) {
    int totalWeight = 0;
    for (const auto& edge : mst) {
        totalWeight += edge.weight;
    }
    return totalWeight;
}

int main() {
    int n = 5; // Number of nodes in the power grid
    vector<Edge> edges = {
        {0, 1, 10}, {0, 2, 6}, {0, 3, 5},
        {1, 3, 15}, {2, 3, 4}, {1, 2, 7}
    };
    
    // Define valid weight range
    int minWeight = 1;
    int maxWeight = 20;

    try {
        // Validate edge weights
        validateEdgeWeights(edges, minWeight, maxWeight);
    } catch (const invalid_argument& e) {
        cerr << "Error: " << e.what() << endl;
        return 1; // Exit if validation fails
    }
    
    // Create adjacency list representation of the graph
    vector<vector<pair<int, int>>> graph(n);
    for (const auto& edge : edges) {
        graph[edge.src].push_back({edge.dest, edge.weight});
        graph[edge.dest].push_back({edge.src, edge.weight});
    }

    // Compute the MST using Prim's algorithm
    vector<Edge> mst = primMST(graph, n);
    
    // Add redundancy to handle up to 2 edge failures
    int k = 2;
    vector<Edge> redundantEdges = addRedundancy(edges, mst, n, k);

    cout << "Edges in the Minimum Spanning Tree (Prim's algorithm):" << endl;
    for (const auto& edge : mst) {
        cout << edge.src << " - " << edge.dest << ": " << edge.weight << endl;
    }

    cout << "Total Weight of the Minimum Spanning Tree: " << calculateMSTWeight(mst) << endl;

    cout << "Additional Redundant Edges for Reliability:" << endl;
    for (const auto& edge : redundantEdges) {
        cout << edge.src << " - " << edge.dest << ": " << edge.weight << endl;
    }

    return 0;
}

File: Codes/Final_Codes_Combined/prims.py
from typing import List, Tuple
import heapq

# Edge structure
class Edge:
    def __init__(self, src: int, dest: int, weight: int):
        self.src = src
        self.dest = dest
        self.weight = weight

# Function to compute the MST using Prim's algorithm
def primMST(edges: List[Edge], n: int) -> List[Edge]:
    # Create an adjacency list for the graph
    adj = {i: [] for i in range(n)}
    for edge in edges:
        adj[edge.src].append((edge.weight, edge.dest))
        adj[edge.dest].append((edge.weight, edge.src))  # Undirected graph

    # Priority queue to select minimum weight edge
    pq = [(0, 0)]  # (weight, node) starting from node 0
    in_mst = [False] * n  # Track visited nodes
    mst = []  # Store edges in the MST
    total_edges = 0

    while pq and total_edges < n - 1:
        weight, node = heapq.heappop(pq)
        if in_mst[node]:
            continue

        in_mst[node] = True
        for neighbor_weight, neighbor in adj[node]:
            if not in_mst[neighbor]:
                heapq.heappush(pq, (neighbor_weight, neighbor))
                mst.append(Edge(node, neighbor, neighbor_weight))
                total_edges += 1

    return mst[:total_edges]  # Return only valid MST edges

# Function to calculate the total weight of the MST
def calculateMSTWeight(mst: List[Edge]) -> int:
    return sum(edge.weight for edge in mst)

# Function to add redundancy
def addRedundancy(edges: List[Edge], mst: List[Edge], n: int, k: int) -> List[Edge]:
    mst_edges_set = {(min(edge.src, edge.dest), max(edge.src, edge.dest)) for edge in mst}
    redundant_edges = [edge for edge in edges if (min(edge.src, edge.dest), max(edge.src, edge.dest)) not in mst_edges_set]

    # Sort redundant edges by weight
    redundant_edges.sort(key=lambda edge: edge.weight)
    return redundant_edges[:k]

# Function to process input and compute MSTs
def processInputFile(filename: str):
    with open(filename, "r") as file:
        lines = file.readlines()

    idx = 0
    graph_count = 1

    while idx < len(lines):
        # Skip empty lines
        if not lines[idx].strip():
            idx += 1
            continue

        # Read n (nodes) and k (redundant edges to add)
        try:
            n, k = map(int, lines[idx].strip().split())
            idx += 1
        except ValueError:
            print(f"Invalid graph header format: {lines[idx].strip()}")
            idx += 1
            continue

        # Read edges until the next header or end of file
        edges = []
        while idx < len(lines) and lines[idx].strip():
            try:
                src, dest, weight = map(int, lines[idx].strip().split())
                edges.append(Edge(src, dest, weight))
                idx += 1
            except ValueError:
                print(f"Invalid edge format: {lines[idx].strip()}")
                idx += 1
                continue

        # Compute MST
        mst = primMST(edges, n)
        total_weight = calculateMSTWeight(mst)

        # Add redundant edges
        redundant_edges = addRedundancy(edges, mst, n, k)

        # Display results
        print(f"Graph {graph_count}:")
        print("Edges in the Minimum Spanning Tree:")
        for edge in mst:
            print(f"{edge.src} - {edge.dest}: {edge.weight}")
        print(f"Total Weight of the Minimum Spanning Tree: {total_weight}")
        print("Additional Redundant Edges for Reliability:")
        for edge in redundant_edges:
            print(f"{edge.src} - {edge.dest}: {edge.weight}")
        print()
        graph_count += 1

# Main function
# if __name__ == "__main__":
#     input_file = "input.txt"  # Specify the input file name
#     try:
#         processInputFile(input_file)
#     except FileNotFoundError:
#         print(f"Error: The file '{input_file}' does not exist in the current directory.")
